src/backend/resources.rs

```diff
@@ -0,0 +1,228 @@
+        let cell_ref = cell.borrow();
+        let primitives = vec![
+            "std_reg",
+            "std_mem_d1",
+            "std_mem_d2",
+            "std_mem_d3",
+            "std_mem_d4",
+            "seq_mem_d1",
+            "seq_mem_d2",
+            "seq_mem_d3",
+            "seq_mem_d4",
+        ];
+        match &cell_ref.prototype {
+            CellType::Primitive {
+                name,
+                param_binding,
+                ..
+            } => {
+                if primitives.contains(&name.id.as_str()) {
+                    *count_map
+                        .entry((*name, (**param_binding).clone()))
+                        .or_insert(0) += 1;
+                }
+            }
+            CellType::Component { name } => {
+                let component = ctx
+                    .components
+                    .iter()
+                    .find(|comp| comp.name == name)
+                    .unwrap();
+                gen_count_map(ctx, component, count_map);
+            }
+            _ => (),
+        }
+    }
+}
+
+/// Writes a CSV to stdout with primitive count information
+/// generated by `gen_count_map`.
+/// TODO (priya): fix the output format and have it output to an actual file!
+fn write_csv(count_map: HashMap<(Id, Binding), u32>, _file: &mut OutputFile) {
+    let mut wtr = csv::Writer::from_writer(io::stdout());
+    let header = vec!["Primitive", "Count", "Parameters"];
+    wtr.write_record(header).ok();
+    wtr.write_record(None::<&[u8]>).ok();
+    for ((name, params), count) in count_map {
+        let mut result = String::new();
+        result.push_str(name.id.as_str());
+        result.push(',');
+        result.push_str(count.to_string().as_str());
+        for (id, val) in params {
+            result.push(',');
+            result.push_str(id.id.as_str());
+            result.push(',');
+            result.push_str(val.to_string().as_str());
+        }
+        result.push('\n');
+        let ret: Vec<&str> = result.split(',').collect();
+        wtr.write_record(ret).ok();
+    }
+    wtr.flush().ok();
+}
+
+// Prints the estimated size (in bits) of the generated hardware along with a breakdown
+// of which primitives contributed to the total number.
+// TODO (priya): Add other primitives (need to discuss what reasonable size estimates are for them)
+fn estimated_size(count_map: HashMap<(Id, Binding), u32>) {
+    let mut estimated_size: u64 = 0;
+    let out = &mut (Box::new(io::stdout()) as Box<dyn Write>);
+    writeln!(out).ok();
+    output("Summary of primitives:".to_string(), out);
+    for ((name, params), count) in count_map {
+        match name.as_ref() {
+            "std_reg" => {
+                estimated_size += (count as u64) * (params[0].1);
+                let str = format!(
+                    "{} {} primitive(s) with a bitwidth of {}.",
+                    count, name, params[0].1
+                );
+                output(str, out);
```

Comment:
the `write!` and `writeln!` macros can do this same formatting so this can be rolled into a single macro invocation rather than saving the local variable